peps/pep-0734.rst

```diff
@@ -0,0 +1,396 @@
+  the ``__main__`` module, neither before nor after, so each
+  successive call picks up where the last one left off.  This can
+  be useful for running some code to initialize an interpreter
+  (e.g. with imports) before later performing some repeated task.
+
+Comparison with builtins.exec()
+-------------------------------
+
+``Interpreter.exec()`` is essentially the same as the builtin
+``exec()``, except it targets a different interpreter, using that
+interpreter's isolated state.
+
+The builtin ``exec()`` runs in the current OS thread and pauses
+whatever was running there, which resumes when ``exec()`` finishes.
+No other threads are affected.  (To avoid pausing the current thread,
+run ``exec()`` in a ``threading.Thread``.)
+
+``Interpreter.exec()`` works the same way.
+
+The builtin ``exec()`` executes against a namespace, by default the
+``__dict__`` of the current module (i.e. ``globals()``).
+It uses that namespace as-is and does not clear it before or after.
+
+``Interpreter.exec()`` works the same way.
+
+...where the "current" module is always the ``__main__`` module
+of the target interpreter.  This is the same as how it works
+for code run from the Python command-line (e.g. ``-m``) or the REPL.
+
+The builtin ``exec()`` discards any object returned from the
+executed code.
+
+``Interpreter.exec()`` works the same way.
+
+The builtin ``exec()`` propagates any uncaught exception from the code
+it ran.  The exception is raised from the ``exec()`` call in the
+thread that originally called ``exec()``.
+
+``Interpreter.exec()`` works the same way.
+
+...with one slight difference.  Rather than propagate the uncaught
+exception directly, we raise an ``interpreters.RunFailedError``
+with a snapshot of the uncaught exception (including its traceback)
+as the ``__cause__``.  Directly raising (a proxy of) the exception
+is problematic since it's harder to distinguish between an error
+in the ``Interpreter.exec()`` call and an uncaught exception
+from the subinterpreter.
+
+API: Communicating Between Interpreters
+---------------------------------------
+
+The module introduces a basic communication mechanism called "channels".
+They are based on `CSP`_, as is Go's concurrency model (loosely).
+Channels are like pipes: FIFO queues with distinct send/receive ends.
+They are designed to work safely between isolated interpreters.
+
+.. _CSP:
+   https://en.wikipedia.org/wiki/Communicating_sequential_processes
+
+For now, only objects that are specifically supported for passing
+between interpreters may be sent through a channel.
+See `Shareable Objects`_.
+
+The module's top-level API for this new mechanism:
+
++----------------------------------------------------+-----------------------+
+| signature                                          | description           |
++====================================================+=======================+
+| ``create_channel() -> (RecvChannel, SendChannel)`` | Create a new channel. |
```

Comment:
`RecvChannel` is the receiving end and `SendChannel` is the sending end.  Perhaps it would help if they were something like `RecvChannelEnd` and `SendChannelEnd`?

Reply:
Hm, those don't delight me, but I can't think of something better either. I wonder if something using sender/receiver or reader/writer could be used?

The key insight that I eventually reached was that this returns two ends of a queue, and the API is a bit like `r, w = os.pipe()`. Including the part where you need to "give" the other party the other end of a channel you created. Which you can do using either `set_main_attr()` or by sending it over an already established channel. And that's why channels are "shareable" objects (rather uniquely, since all the other shareable types are basic data types).

Reply:
Honestly, I'm probably going to drop the whole "distinct ends" thing, have one class, and just call it `Queue`, in the spirit of the minimal API I'm aiming for.  For that matter, I suppose we could put it in the `queue` module (something like `CrossInterpreterQueue`), but it probably still belongs in `interpreters`.  We could add directional channels later, if worth doing.

My only reservation is that single-directional channel ends often makes code easier to reason about and harder to do the wrong thing.  A bi-directional queue object has the opposite effect.  At least, that was my experience with Go, which has both directional and bi-directional channels.  A bi-directional channel was never necessary but it was easy to use one without thinking (at the expense of extra cognitive overhead later), since that's what you get when you create a channel (e.g. `var foo = make(chan int)`).  Having only directional channels from the start eliminates the problem, which is part of why I've been proposing `RecvChannel` and `SendChannel`.  In some ways it's better to not introduce the less desirable thing in the first place, rather than try to discourage its use (or remove it) later.

However, given your reaction, I'm inclined to think it isn't worth having the distinct ends.  Furthermore, I don't recall anyone bringing up concerns with Python's `queue.Queue` like I've experienced with Go's bi-directional channel.  So it may very well be a Go-only issue and one I needn't worry about with the `interpreters` module.

Reply:
A single Queue class sounds good, just don't put it in queue.py (it's not really related except at a high level. For example, asyncio has its own queue submodule.

I wouldn't worry too much about not having distinct read/write ends -- we live with that for file I/O too, for example. I presume that using a Queue, an interpreter could possibly send to itself, which seems fine as an edge case.

Is there something about queue capacity?

Reply:
> Is there something about queue capacity?

I didn't see a need to worry about anything but unbounded for now.

Reply:
Okay, but then why do we have both `send()` and `send_nowait()`? If the queue is unbounded there's never a need to block in `send()`, is there? How is the difference implemented?

(I have no beef with `recv_nowait()` -- if the queue is empty, one blocks, the other doesn't. But in queue.py, `put_nowait()` exists to avoid blocking when the queue is *full*.)

Reply:
In that context you can think of the queue as having a max size of 0 then.

FWIW, I've found it useful to know when the object is actually popped off the queue, as a synchronization point between two interpreters.

Reply:
Oh wait, the updated description makes it clear that queues have max capacity zero. That's surprising given that previously you described it as similar to `os.pipe()`, but it is a consistent model, so I withdraw my objections.

Now I vaguely wonder how we'd implement a queue *with* buffering on top of this, but that's a straightforward programming problem.